src/Datadog.Trace.ClrProfiler.Managed/Integrations/RabbitMQIntegration.cs

```diff
@@ -0,0 +1,900 @@
+        /// <param name="moduleVersionPtr">A pointer to the module version GUID.</param>
+        [InterceptMethod(
+            TargetAssembly = RabbitMQAssembly,
+            TargetType = RabbitMQImplModelBase,
+            TargetMethod = "_Private_BasicPublish",
+            TargetSignatureTypes = new[] { ClrNames.Void, ClrNames.String, ClrNames.String, ClrNames.Bool, ClrNames.Ignore, ClrNames.Ignore },
+            TargetMinimumVersion = Major3Minor6Patch9,
+            TargetMaximumVersion = Major5)]
+        public static void BasicPublish(
+            object model,
+            string exchange,
+            string routingKey,
+            bool mandatory,
+            object basicProperties,
+            byte[] body,
+            int opCode,
+            int mdToken,
+            long moduleVersionPtr)
+        {
+            if (model == null) { throw new ArgumentNullException(nameof(model)); }
+
+            const string methodName = "_Private_BasicPublish";
+            const string command = "basic.publish";
+            Action<object, string, string, bool, object, byte[]> instrumentedMethod;
+            var modelType = model.GetType();
+
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Action<object, string, string, bool, object, byte[]>>
+                       .Start(moduleVersionPtr, mdToken, opCode, methodName)
+                       .WithConcreteType(modelType)
+                       .WithParameters(exchange, routingKey, mandatory, basicProperties, body)
+                       .WithNamespaceAndNameFilters(ClrNames.Void, ClrNames.String, ClrNames.String, ClrNames.Bool, ClrNames.Ignore, ClrNames.Ignore)
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorRetrievingMethod(
+                    exception: ex,
+                    moduleVersionPointer: moduleVersionPtr,
+                    mdToken: mdToken,
+                    opCode: opCode,
+                    instrumentedType: RabbitMQImplModelBase,
+                    methodName: methodName,
+                    instanceType: modelType.AssemblyQualifiedName);
+                throw;
+            }
+
+            RabbitMQTags tags = null;
+            using (var scope = CreateScope(Tracer.Instance, out tags, command, spanKind: SpanKinds.Producer, exchange: exchange, routingKey: routingKey))
+            {
+                if (scope != null)
+                {
+                    string exchangeDisplayName = string.IsNullOrEmpty(exchange) ? "<default>" : exchange;
+                    string routingKeyDisplayName = string.IsNullOrEmpty(routingKey) ? "<all>" : routingKey.StartsWith("amq.gen-") ? "<generated>" : routingKey;
+                    scope.Span.ResourceName = $"{command} {exchangeDisplayName} -> {routingKeyDisplayName}";
+
+                    tags.MessageSize = body.Length.ToString();
```

Comment:
You're checking if `tags == null` in other places. Should we that that here as well? (and a few lines below)

Reply:
Good idea, added a null check on `tags` and on `body` in 92957e00598a231e631db4348ee402f93ff6bdfc. Missed one so I also added a missing null check in 3ecaf40b6702d7149298feef9b014b685a0ceddc